contracts/Parameterizer.sol

```diff
@@ -1,96 +1,323 @@
+      name: _name,
+      owner: msg.sender,
+      processBy: now + get("pApplyStageLen") + get("pCommitStageLen") +
+        get("pRevealStageLen") + PROCESSBY,
+      value: _value
+    });
+
+    _ReparameterizationProposal(msg.sender, _name, _value, propID);
+    return propID;
+  }
+
+  /**
+  @notice challenge the provided proposal ID, and put tokens at stake to do so.
+  @param _propID the proposal ID to challenge
+  */
+  function challengeReparameterization(bytes32 _propID) public returns (uint challengeID) {
+    ParamProposal memory prop = proposalMap[_propID];
+    uint deposit = get("pMinDeposit");
+
+    require(propExists(_propID) && prop.challengeID == 0); 
+
+    //take tokens from challenger
+    require(token.transferFrom(msg.sender, this, deposit));
+    //start poll
+    uint pollID = voting.startPoll(
+      get("pVoteQuorum"),
+      get("pCommitStageLen"),
+      get("pRevealStageLen")
+    );
+
+    challengeMap[pollID] = Challenge({
+      challenger: msg.sender,
+      rewardPool: ((100 - get("pDispensationPct")) * deposit) / 100, 
+      stake: deposit,
+      resolved: false,
+      totalTokens: 0
+    });
+
+    proposalMap[_propID].challengeID = pollID;       // update listing to store most recent challenge
+
+    _NewChallenge(msg.sender, _propID, pollID);
+    return pollID;
+  }
+
+  /**
+  @notice for the provided proposal ID, set it, resolve its challenge, or delete it depending on whether it can be set, has a challenge which can be resolved, or if its "process by" date has passed
+  @param _propID the proposal ID to make a determination and state transition for
+  */
+  function processProposal(bytes32 _propID) public {
+    ParamProposal storage prop = proposalMap[_propID];
+
+    if (canBeSet(_propID)) {
+      set(prop.name, prop.value);
+    } else if (challengeCanBeResolved(_propID)) {
+      resolveChallenge(_propID);
+    } else if (now > prop.processBy) {
+    } else {
+      revert();
+    }
+
+    delete proposalMap[_propID];
+  }
+
+  /**
+  @notice claim the tokens owed for the msg.sender in the provided challenge
+  @param _challengeID the challenge ID to claim tokens for
+  @param _salt the salt used to vote in the challenge being withdrawn for
+  */
+  function claimReward(uint _challengeID, uint _salt) public {
+    // ensure voter has not already claimed tokens and challenge results have been processed
+    require(tokenClaims[_challengeID][msg.sender] == false);
+    require(challengeMap[_challengeID].resolved = true);
+
+    uint voterTokens = voting.getNumPassingTokens(msg.sender, _challengeID, _salt);
+    uint reward = calculateVoterReward(msg.sender, _challengeID, _salt);
+
+    // subtract voter's information to preserve the participation ratios of other voters
+    // compared to the remaining pool of rewards
+    challengeMap[_challengeID].totalTokens -= voterTokens;
+    challengeMap[_challengeID].rewardPool -= reward;
+
+    require(token.transfer(msg.sender, reward));
```

Comment:
as a design patter, should the state transition be done before the transfer (even though re-entrancy may not possible here).